software/spmd/bsg_attr_noalias_ex/saxpy-c.c

```diff
@@ -0,0 +1,132 @@
+#include <bsg_manycore.h>
+#include "saxpy-c.h"
+
+/*
+ * saxpy_c result: Without bsg_attr_noalias, the compiler cannot guarantee that
+ * stores to C do not affect A and B, so load/store instructions are not
+ * reordered to overlap loop iterations and hide load latency.
+ *
+ * The disassembly and compiler flags are shown in snippets/saxpy_c.md
+ */
+void saxpy_c(float  *  A, float  *  B, float *C, float alpha) {
+        bsg_unroll(4)
+        for(int i = 0;  i < N_ELS; ++i) {
+                C[i] = alpha * A[i] + B[i];
+        }
+}
+
+
+/*
+ * saxpy_c_noalias: With bsg_attr_noalias, the code "guarantees" that stores to
+ * C do not affect A and B so load/store instructions can be reordered because
+ * the pointers do not alias.
+ *
+ * The disassembly and compiler flags are shown in snippets/saxpy_c_noalias.md
+ *
+ */
+void saxpy_c_noalias(float * bsg_attr_noalias A, float * bsg_attr_noalias B, float * bsg_attr_noalias C, float alpha) {
+        bsg_unroll(4)
+        for(int i = 0;  i < N_ELS; ++i) {
+                C[i] = alpha * A[i] + B[i];
+        }
+}
+
+
+/*
+ * Use caution when applying bsg_attr_noalias to multi-dimensional
+ * pointers. In this example LLVM produces worse code than GCC.
+
+ * This is a known issue in LLVM with a patch under review (as of 8/20):
+ * Google: restrict support in llvm
+ * LLVM Patch: https://reviews.llvm.org/D69542
```

Comment:
Just fyi, I had tried this patch but found that noalias information was not being passed onto the llvm backend. The patch author's main goal has been to enable vectorization, which needs similar guarantees, on the data pointed by pointer, to that of our use case. https://lists.llvm.org/pipermail/llvm-dev/2020-June/142684.html